File: src/main/scala/ApiProtocol.scala
case class Problem(
  id: String,
  size: Int,
  operators: Array[String],
  solved: Option[Boolean],
  timeLeft: Option[Int])

case class EvalRequest(
  id: Option[String],
  program: Option[String] = None,
  arguments: Array[String])

case class EvalResponse(
  status: String,
  outputs: Option[Array[String]],
  message: Option[String])

case class Guess(
  id: String,
  program: String)

case class GuessResponse(
  status: String,
  values: Option[Array[String]],
  message: Option[String],
  lightning: Option[Boolean])

case class TrainRequest(
  size: Option[Int] = None,
  operators: Option[Array[String]] = None)

case class TrainingProblem(
  challenge: String,
  id: String,
  size: Int,
  operators: Array[String])

case class Window(
  resetsIn: Int,
  amount: Int,
  limit: Int)

case class Status(
  easyChairId: String,
  contestScore: Int,
  lightningScore: Int,
  trainingScore: Int,
  mismatches: Int,
  numRequests: Int,
  requestWindow: Window,
  cpuWindow: Window,
  cpuTotalTime: Int)

File: fsharp/icfpc2013/icfpc2013/ApiProtocol.fs
module ApiProtocol

type TrainRequest = { 
    size: int option
    operators: array<string> option
    }

type TrainingProblem = {
     challenge: string
     id: string
     size: int
     operators: array<string>
     }

type EvalRequest = {
    id: string option
    program: string option
    arguments: array<string>
}

type EvalResponse = {
    status: string 
    outputs: array<string> option  
    message: string option
}

type Guess = {
    id: string
    program: string
    }

type GuessResponse = {
    status: string
    values: array<string> option
    message: string option
    lightning: bool option
    }

// partial description
type Status = {
    easyChairId: string
    contestScore: int
    lightningScore: int
    trainingScore: int
    mismatches: int
    numRequests: int
    }

type Problem = {
    id: string
    size: int
    operators: array<string>
    solved: bool option
    timeLeft: int option   
}

